lib/aggregators/BaseAggregator.ts

```diff
@@ -43,9 +49,59 @@ export abstract class BaseAggregator<State> {
     return undefined;
   }
 
-  abstract init(start: RDF.Term): State;
+  protected abstract subInit(start: RDF.Term | undefined): SubState;
+  protected abstract subPut(state: SubState, bindings: RDF.Term | undefined): SubState;
+  protected abstract subResult(state: SubState): RDF.Term;
+
+  public init(start: RDF.Term | undefined, variable: string): IBaseState & { sub: SubState } {
+    const subState: SubState = this.subInit(start);
+    return {
+      sub: subState,
+      seen: this.addSeen({}, start, variable),
+    };
+  }
+
+  public put(state: IBaseState & { sub: SubState }, bindings: RDF.Term | undefined, variable: string):
+  IBaseState & { sub: SubState } {
+    if (this.canSkip(state, bindings, variable)) {
+      return state;
+    }
+    const sub: SubState = this.subPut(state.sub, bindings);
+    return { sub, seen: this.addSeen(state.seen, bindings, variable) };
+  }
+
+  public result(state: IBaseState & { sub: SubState }): RDF.Term {
+    return this.subResult(state.sub);
+  }
+
+  protected canSkip(state: IBaseState, term: RDF.Term | undefined, variable: string): boolean {
+    return this.distinct && (state.seen[variable] || []).some((t: RDF.Term) => t.equals(term));
+  }
+
+  protected addSeen(seen: Seen, term: RDF.Term | undefined,
+    variable: string): Seen {
+    return this.distinct ?
+      {
+        ...seen,
```

Comment:
This operation may get a bit too expensive due to object re-recreation and copying.

I suggest to make use of a mutable `Map<string, Set<string>>`, which maps variables to a set of stringified (using `rdf-string`) terms.
The stringified terms can then just be checked using `has` instead of iteration.

Reply:
Perfect! I we're working with mutable types. I think we should stop passing the state as arguments to for example the `put` function.